src/main/scala/org/ergoplatform/nodeView/wallet/TrackedBox.scala

```diff
@@ -0,0 +1,250 @@
+package org.ergoplatform.nodeView.wallet
+
+import org.ergoplatform.ErgoBox
+import org.ergoplatform.modifiers.mempool.ErgoTransaction
+import org.ergoplatform.nodeView.history.ErgoHistory.Height
+import org.ergoplatform.nodeView.wallet.BoxCertainty.Certain
+import org.ergoplatform.nodeView.wallet.OnchainStatus.{Offchain, Onchain}
+import org.ergoplatform.nodeView.wallet.SpendingStatus.{Spent, Unspent}
+import org.ergoplatform.settings.Algos
+import scorex.core.{ModifierId, bytesToId}
+import scorex.core.utils.ScorexLogging
+
+/**
+  * A generic interface for a box tracked by a wallet. A TrackedBox instantiation contains box itself as well as
+  * its state (e.g. spent or not, confirmed or not etc).
+  */
+sealed trait TrackedBox extends ScorexLogging {
+
+  /**
+    * Whether the box is spent or not
+    */
+  def spendingStatus: SpendingStatus
+
+  /**
+    * Whether the box is confirmed or not
+    */
+  def onchainStatus: OnchainStatus
+
+  /**
+    * Whether the box is definitely belongs to the user or not
+    */
+  def certainty: BoxCertainty
+
+  final def spent: Boolean = spendingStatus.spent
+  final def onchain: Boolean = onchainStatus.onchain
+  final def certain: Boolean = certainty.certain
+
+  /**
+    * Transaction created the box
+    */
+  def creationTx: ErgoTransaction
+
+  /**
+    * Output index in the transaction
+    */
+  def creationOutIndex: Short
+
+  val box: ErgoBox
+
+  lazy val boxId: ModifierId = bytesToId(box.id)
+
+  def value: Long = box.value
+
+  lazy val assets: Map[ModifierId, Long] = box.additionalTokens.map { case (id, amt) =>
+    bytesToId(id) -> amt
+  }.toMap
+
+  /**
+    * Register this box in a wallet storage
+    */
+  def register(storage: WalletStorage): Unit = storage.putToRegistry(this)
```

Comment:
Why such a syntax? Our regular syntax is `storage.put(object)` rather than `object.register(storage)`